### PR TITLE
[feat] pack the GenRec training forward through FlashAttention-2

### PR DESCRIPTION
**Pull request by WhiteSwan1**
GenRec padded every variable-length prompt before the teacher-forced forward, so every row paid for the longest row in its batch. This runs that path on packed embeddings through the FlashAttention-2 varlen kernel instead: one concatenated stream, explicit `cu_seq_lens` boundaries, per-row `position_ids`, and response-only logits gathered straight out of the packed layout.

## Measured

On 1× H20, `GenRecCausalLMModel` over Qwen2.5-0.5B:

| batch 20, real length | SDPA (the path this replaces) | FLASH_ATTENTION_2 |
| --- | ---: | ---: |
| step time | 448.8 ms | **233.4 ms — 1.92×** |
| peak memory | 31,515 MiB | **20,147 MiB — −36%** |

The margin grows with sequence length: **2.55×** at 2048 tokens, where SDPA peaks at 76.8 GB on a 96 GB card against FA2's 41.1 GB.

Worth reading the breakdown before optimising further: after packing, attention is only **7.0%** of the step at production length, and the linear layers are 60% running at ~90% of H20 BF16 peak. The win here is mostly *not paying for padding* — 43% of tokens at batch 20 — rather than a faster attention kernel.

## Choosing the kernel

`GenRecModelConfig.common.attn_kernel` picks the backend, carrying the transformers name straight through to `attn_implementation`:

```proto
optional string attn_kernel = 6 [default = "sdpa"];
```

`sdpa` is the default because it is the backend whose dependency always ships — building a GenRec model then needs neither a GPU nor the `flash_attn` wheel, which keeps the CPU lane and export working. Training runs opt into `flash_attention_2` explicitly. There is no `AUTO` and no silent fallback: a misconfiguration surfaces as the library's own error rather than a quietly slower path.

**Each kernel gets the layout it can actually serve.** Only the varlen kernel reads `cu_seq_lens`; every other kernel rebuilds the boundaries as a dense `(1, 1, T, T)` mask, which costs `(ΣL)²` where left-padding costs `Σ(B·L²)`. So `_forward` branches — packed for `flash_attention_2`, left-padded for everything else. The supervised window is the same absolute positions either way, so the labels are built once and only the logits differ. Measured at 8 rows of 128: packed is 1,048,576 score-pairs behind a 1 MiB mask, padded is 131,072 behind no mask at all.

`flash_attn` is declared in `requirements/cu126|cu129|cu130.txt` beside the other self-hosted wheels — three lines each, one per supported Python (3.10/3.11/3.12), matching the coverage `dynamicemb`, `faiss` and `fbgemm_gpu_hstu` already have.

## Test plan

- Unit coverage for the packed varlen metadata, the suffix window and its labels, response-only loss and gradients, the short-row guard, and backbone construction under both kernels.
- `_RowIsolationCase` asserts the invariant that matters — rewrite row 0 and row 1's logits must stay bit-identical — and runs under **both** kernels: the sdpa arm on the CPU lane on every PR, the flash arm on the GPU lane.
- Verified the two branches agree to `5.96e-08` with identical labels.
- GPU coverage for qwen2/qwen3 packed-vs-solo forwards, projected-slot gradients under BF16/FP16 and under fp32 masters with BF16 autocast, and ragged beam decoding.

## Known gaps

- **The wheel is locked to the torch minor version** through libtorch's C++ ABI, not just to CUDA: it links 70 `c10::`/`at::` symbols against unversioned sonames, so a torch bump without a rebuild fails at *import*, not at install. `Requires-Dist: torch` carries no bound, so pip cannot catch the mismatch.
- **cu126 has no sm100/sm120 build**, unlike cu129 and cu130. Blackwell on a CUDA 12.6 install falls back to SDPA.

FA3 was measured too and is **not** proposed here: +2.5% at production length, no memory saving, and no wheel — it needs a from-source Hopper build against the same torch minor ABI.

### CODE REVIEW

File: tzrec/models/genrec_causal_lm_model.py [hunk 1]
```diff
@@ -143,17 +143,94 @@ def _forward(
...
         """
-        padded, mask, labels = self._left_pad_packed_inputs(
-            embeds, batch, build_labels=True
+        infos = batch.additional_infos
+        cu_seqlens = infos[CU_SEQLENS]
+        lengths = torch.diff(cu_seqlens)
+        suffix = cast(int, self._prompt.prompt_plan.logits_suffix_len)
+        # a row shorter than the window reads the row before it
+        if bool((lengths < suffix).any()):
+            raise ValueError(
```
[the remark this answers is not in the document]
Comment (WhiteSwan1) [reply tzrec/models/genrec_causal_lm_model.py]: Fixed in `53ab60b`: moved into `PromptAssembler`, which runs on the host, so the sync
is gone. Serving uses `include_response=False` and needs no guard.

File: tzrec/models/genrec_causal_lm_model.py [hunk 2]
```diff
@@ -176,17 +253,15 @@ def _left_pad_packed_inputs(
...
             embeds: packed embeddings, ``(total_tokens, hidden)``.
             batch: carries the packed prompt metadata.
-            build_labels: whether to build response-only training labels.
 
         Returns:
-            Padded embeddings, attention mask and optional labels.
+            Padded embeddings and attention mask.
         """
         infos = batch.additional_infos
         cu_seqlens = infos[CU_SEQLENS]
```
[the remark this answers is not in the document]
Comment (WhiteSwan1) [reply tzrec/models/genrec_causal_lm_model.py:271]: Fixed in `53ab60b`. `5043e15` then renamed it `_left_pad` — the old name led with its
input, so call sites read `padded = ..._packed_inputs(...)`.

File: tzrec/models/genrec_causal_lm_model.py [hunk 3]
```diff
@@ -143,17 +143,94 @@ def _forward(
...
+
+        Args:
+            embeds: the assembled prompt embeddings, packed.
+            batch: carries the row boundaries and the collator's width.
+            suffix: width of the supervised window.
+
+        Returns:
+            ``(rows, suffix, vocab)`` logits over the response window.
+        """
+        padded, mask = self._left_pad_packed_inputs(embeds, batch)
```
[the remark this answers is not in the document]
Comment (WhiteSwan1) [reply tzrec/models/genrec_causal_lm_model.py]: Stale: `233a152` deleted that exclusion — `grep -rn CUDNN_ATTENTION tzrec/` is empty.
That removal is what made the docstring true.

File: tzrec/models/genrec_causal_lm_model_test.py [hunk 4]
```diff
@@ -87,47 +224,143 @@ def setUp(self) -> None:
...
+        )
+    )
+    return batch
+
+
+class _RowIsolationCase:
+    """Every row must read exactly as it does alone, whatever the layout.
+
+    The failure this guards against is a row attending into the row before it,
+    which the two kernels avoid by different means: the varlen kernel is handed
```
[the remark this answers is not in the document]
Comment (WhiteSwan1) [reply tzrec/models/genrec_causal_lm_model_test.py]: Added `test_the_packed_and_padded_arms_agree`: one seed, SDPA vs FLASH_ATTENTION_2, both
BF16 on CUDA. The second bullet's reason went with the cuDNN removal (3).

File: tzrec/utils/test_util.py [hunk 5]
```diff
@@ -31,6 +31,9 @@
 from tzrec.protos.models.genrec_model_pb2 import GenRecModelConfig
 from tzrec.protos.prompt_pb2 import PromptSlot
 from tzrec.utils.export_util import split_model
+
+if TYPE_CHECKING:
+    from tzrec.features.feature import BaseFeature
```
[the remark this answers is not in the document]
Comment (WhiteSwan1) [reply tzrec/utils/test_util.py:36]: False positive — isort sorts contiguous blocks and the `if` terminates one.
`ruff check --select I` passes, and Code Style CI is green.

File: tzrec/models/genrec_causal_lm_model.py [hunk 6]
```diff
@@ -143,17 +143,94 @@ def _forward(
...
+        suffix = cast(int, self._prompt.prompt_plan.logits_suffix_len)
+        # a row shorter than the window reads the row before it
+        if bool((lengths < suffix).any()):
+            raise ValueError(
+                f"{type(self).__name__}: every assembled sample must be at "
+                f"least logits_suffix_len ({suffix}) tokens long. Drop the "
+                f"rows whose prompt features are all empty, or give the "
+                f"template static text."
+            )
+        suffix_offsets = torch.arange(-suffix, 0, device=embeds.device)
```
[the remark this answers is not in the document]
Comment (WhiteSwan1) [reply tzrec/models/genrec_causal_lm_model.py:149]: Fixed in `53ab60b`, but neither way you proposed. The guard is load-bearing: a
prompt-less row's first label is predicted from outside the row, unmasked. It also isn't
packed-only — on SDPA that logit lands on a pad column. The defect was the predicate,
`L >= response_max + 1` where `L >= response_length + 1` suffices. `PromptAssembler` now
zeroes those rows' `response_lengths`, filtering them from the loss. No decision needed:
the new predicate is strictly weaker.

File: tzrec/protos/models/genrec_model.proto [hunk 7]
```diff
@@ -24,6 +24,14 @@ message GenRecModelConfig {
...
     // bf16 compute comes from mixed_precision, not from this.
     optional ParamDtype lm_parameter_dtype = 5 [default = FP32];
+
+    // FLASH_ATTENTION_2 is what training should use; it needs the flash_attn
+    // wheel and a GPU. SDPA runs anywhere.
+    enum AttnKernel {
+        SDPA = 0;
+        FLASH_ATTENTION_2 = 1;
+    }
+    optional AttnKernel attn_kernel = 6 [default = SDPA];
```
Comment (tiankongdeguiji) [inline tzrec/protos/models/genrec_model.proto]: Could we use a string for attn_kernel instead of introducing a new protobuf enum here?

This value is essentially passed through to Hugging Face's attn_implementation ("sdpa", "flash_attention_2", etc.), so using the upstream string directly would avoid adding another enum to the proto namespace and also remove the extra _ATTN_KERNEL mapping.

For example:

optional string attn_kernel = 6 [default = "sdpa"];

**Review by tiankongdeguiji — CHANGES_REQUESTED** [review]

**Review by tiankongdeguiji — APPROVED** [review]
